crates/ide/src/ide/assists/inline.rs

```diff
@@ -0,0 +1,152 @@
+use super::{AssistKind, AssistsCtx};
+use crate::def::{AstPtr, ResolveResult};
+use crate::TextEdit;
+use smol_str::{SmolStr, ToSmolStr};
+use syntax::ast::AstNode;
+use syntax::{ast, match_ast, SyntaxNode};
+
+pub(super) fn inline_from_reference(ctx: &mut AssistsCtx<'_>) -> Option<()> {
+    let file_id = ctx.frange.file_id;
+    let parse = ctx.db.parse(file_id);
+
+    let covering_node = ctx.covering_node::<ast::Ref>()?;
+    let covering_token = covering_node.token()?;
+    let ptr = AstPtr::new(covering_node.syntax());
+
+    let name_res = ctx.db.name_resolution(file_id);
+    let source_map = ctx.db.source_map(file_id);
+    let expr_id = source_map.expr_for_node(ptr)?;
+
+    let replacement = match name_res.get(expr_id)? {
+        &ResolveResult::Definition(name) => source_map.nodes_for_name(name).find_map(|ptr| {
+            let definition_node = ptr.to_node(&parse.syntax_node());
+            let definition_value = definition_node.ancestors().find_map(|n| {
+                match_ast! {
+                    match n {
+                        ast::AttrpathValue(path_value) => path_value.value(),
+                        _ => None,
+                    }
+                }
+            })?;
+
+            Some(definition_value)
+        }),
+        // Only let-ins are rewritable
+        _ => None,
+    }?;
+    let replacement_text = maybe_parenthesize(&replacement, covering_node.syntax());
+
+    ctx.add(
+        "inline_from_reference",
+        format!("Inline binding `{}`", covering_token.text()),
+        AssistKind::RefactorInline,
+        vec![TextEdit {
+            delete: covering_token.text_range(),
+            insert: replacement_text,
+        }],
+    );
+
+    Some(())
+}
+
+pub(super) fn inline_from_definition(ctx: &mut AssistsCtx<'_>) -> Option<()> {
+    let file_id = ctx.frange.file_id;
+    let name_res = ctx.db.name_resolution(file_id);
+
+    let covering_node = ctx.covering_node::<ast::Attr>()?;
+    let source_map = ctx.db.source_map(file_id);
+    let ptr = AstPtr::new(covering_node.syntax());
+    let name_id = source_map.name_for_node(ptr)?;
+
+    let binding_pathvalue = covering_node
+        .syntax()
+        .ancestors()
+        .find_map(ast::AttrpathValue::cast)?;
+    let binding_definition = binding_pathvalue.value()?;
+
+    let binding_usages = name_res
+        .iter()
+        .filter_map(|(id, res)| match res {
+            &ResolveResult::Definition(def) if def == name_id => source_map.node_for_expr(id),
+            _ => None,
+        })
+        .collect::<Vec<_>>();
+
+    let mut rewrites: Vec<TextEdit> = vec![];
+    let parent = binding_pathvalue.syntax().parent()?;
+    let is_letin = ast::LetIn::cast(parent.clone()).is_some();
+    // FIXME: this would delete into the following expression
+    if is_letin {
+        rewrites.push(TextEdit {
+            delete: parent.text_range(),
+            insert: Default::default(),
+        });
+    };
```

Comment:
`binding_pathvalue.text_range()` (the node of `name = value;`) should be the one to delete, not the whole `LetIn`. Thus the expected diff is `let a = 1; in a` => `let in 1`.

If you mean to delete the empty left-over `let in`, you need to check if the LetIn has a single binding, and to calculate the text range until the end of `in` keyword. `LetIn` expression covers the whole expression including the body expression after `in`.
For simplicity, we can omit this heuristic for now, since we already have another assist to remove empty `let in`.